torch/distributed/_sharded_tensor/ops/equals.py

```diff
@@ -0,0 +1,51 @@
+import torch
+import torch.distributed as dist
+
+def _communicate_result(result, pg):
+    # Gather results from all ranks.
+    if result:
+        result_tensor = torch.ones(1, device=torch.cuda.current_device())
+    else:
+        result_tensor = torch.zeros(1, device=torch.cuda.current_device())
+
+    dist.all_reduce(result_tensor, group=pg)
+
+    expected_result = torch.ones(1, device=torch.cuda.current_device()) * dist.get_world_size(pg)
+
+    return torch.equal(result_tensor, expected_result)
+
+def equal(types, args=(), kwargs=None):
+    from torch.distributed._sharded_tensor import ShardedTensor
+
+    if len(args) != 2:
+        raise ValueError('Expected two arguments for torch.equal')
+
+    result = True
+    st1 = args[0]
+    st2 = args[1]
+    if not(isinstance(st1, ShardedTensor) and isinstance(st2, ShardedTensor)):
+        raise TypeError('Both arguments to torch.equal need to be of type ShardedTensor')
+
+    # Verify same PG
+    if st1._process_group != st2._process_group:
```

Comment:
even if two sts using two pgs, these two pgs have exactly the same set up (i.e. ProcessGroupState is the same), we are still treating it like not equals?

Reply:
I was thinking about this a bit, but if the two PGs are not the same object and we try to compare them it becomes pretty complicated and buggy. The trickiest situation is when two PGs are not the same but overlap on certain ranks. So the ranks where they overlap would consider the PGs same and go down a different path and the ranks that consider them different would go down a different path and result in mismatched `_communicate_result` calls.